### PR TITLE
Support Triton inference server python backend

### PR DESCRIPTION
**Pull request by ispobock**
## Motivation

As mentioned in https://github.com/InternLM/lmdeploy/issues/1309, the Turbomind triton backend is facing some performance issues. So we propose this Python backend. 

We have conducted performance test and OpenCompass evaluation, the results are shown in https://github.com/InternLM/lmdeploy/issues/1309. Its performance is as efficient as the api_server and the evaluation result is also fine.

@lvhan028 @zhyncs 

## Modification

- [x] Add Python backend model file and config file
- [x] Add client and profile script for test
- [x] Add doc

### CODE REVIEW

**Comment by zhyncs** [comment]
LGTM. I have reviewed this PR internally, and everything seems fine in terms of functionality, performance and evaluation. DO you have any suggestions? Thanks. @lvhan028 @lzhangzz @irexyc @grimoire @AllentDan

**Comment by lvhan028** [comment]
linting error
```
AssertionError: Missing files: ['lmdeploy/serve/turbomind/triton_python_backend/__init__.py']
Error: Process completed with exit code 1.
```

**Comment by ispobock** [comment]
> linting error
> 
> ```
> AssertionError: Missing files: ['lmdeploy/serve/turbomind/triton_python_backend/__init__.py']
> Error: Process completed with exit code 1.
> ```

@lvhan028 This is not a Python module, do we still need to add the `__init__.py`?

**Comment by lvhan028** [comment]
I know, but I have no idea how to let it skip the check-module-init action. @RunningLeon any ideas?

**Comment by RunningLeon** [comment]
@zhyncs hi, you can skip the directory in here
https://github.com/InternLM/lmdeploy/blob/451cba29c1af7c322de6ebccebb558eea2351fd5/.github/scripts/check_lmdeploy.py#L21-L22

**Comment by ispobock** [comment]
@zhulinJulia24 could you help review?

**Comment by zhulinJulia24** [comment]
> @zhulinJulia24 could you help review?

Yes, I'm tring on it.

**Comment by zhulinJulia24** [comment]
Should I copy config.json from original model  @ispobock 

```

| lmdeploy_model | 1       | UNAVAILABLE: Internal: FileNotFoundError: [Errno 2] No such file or directory: '/workspace/models/model_repository/lmdeploy_model/1/weights/config.json' |
|                |         |                                                                                                                                                          |
|                |         | At:                                                                                                                                                      |
|                |         |   /opt/lmdeploy/lmdeploy/utils.py(191): get_hf_config_content                                                                                            |
|                |         |   /opt/lmdeploy/lmdeploy/archs.py(107): get_task                                                                                                         |
|                |         |   /opt/lmdeploy/lmdeploy/api.py(64): pipeline                                                                                                            |
|                |         |   /workspace/models/model_repository/lmdeploy_model/1/model.py(36): initialize   
```

**Comment by zhulinJulia24** [comment]
> > Should I copy config.json from original model @ispobock
> > ```
> > 
> > | lmdeploy_model | 1       | UNAVAILABLE: Internal: FileNotFoundError: [Errno 2] No such file or directory: '/workspace/models/model_repository/lmdeploy_model/1/weights/config.json' |
> > |                |         |                                                                                                                                                          |
> > |                |         | At:                                                                                                                                                      |
> > |                |         |   /opt/lmdeploy/lmdeploy/utils.py(191): get_hf_config_content                                                                                            |
> > |                |         |   /opt/lmdeploy/lmdeploy/archs.py(107): get_task                                                                                                         |
> > |                |         |   /opt/lmdeploy/lmdeploy/api.py(64): pipeline                                                                                                            |
> > |                |         |   /workspace/models/model_repository/lmdeploy_model/1/model.py(36): initialize   
> > ```
> 
> I'm wondering how can I get a weights folder? I try to copy it from turbomind workspace, but it's missing tokenizer and config.json

I copy huggingface model into weights, it works

**Comment by ispobock** [comment]
@zhulinJulia24 For huggingface model, you can copy all the files in the model folder to the weights folder. For Turbomind workspace, you can just use the `workspace` and rename it to `weights`. I will update the doc to clarify it.

**Comment by zhulinJulia24** [comment]
@ispobock I use llama2-chat-7b-hf for the test, the performerce near doubled. 
I also use internlm2-chat-20b for test, only one gpu is used, Can you help to have a look?

**Comment by zhyncs** [comment]
> @ispobock I use llama2-chat-7b-hf for the test, the performerce near doubled. I also use internlm2-chat-20b for test, only one gpu is used, Can you help to have a look?

Configure this https://github.com/InternLM/lmdeploy/pull/1329/files#diff-26e1e6c8062566f0e1e5c5ba90c4c029f9c078d86c1ed023763e469c19c6b576R100 if you want to use tp.

**Comment by zhyncs** [comment]
Hi @zhulinJulia24 Under normal circumstances, the performance should be flat or slightly better compared to the API Server, and there should not be a doubling situation. Perhaps you can share your reproduction steps so we can verify it locally.

**Comment by zhulinJulia24** [comment]
> Hi @zhulinJulia24 Under normal circumstances, the performance should be flat or slightly better compared to the API Server, and there should not be a doubling situation. Perhaps you can share your reproduction steps so we can verify it locally.

**1. original triton performence:**
```
--------------------------------------------------
concurrency: 64
elapsed_time: 165.504s

first_token latency(min, max, ave): 0.067s, 1.414s, 0.209s

number of prompt tokens: 244330
number of completion tokens: 240421
token throughput (completion token): 1452.657 token/s
token throughput (prompt + completion token): 2928.932 token/s
RPS (request per second): 6.042 req/s
RPM (request per minute): 362.528 req/min
--------------------------------------------------
--------------------------------------------------
concurrency: 128
elapsed_time: 158.409s

first_token latency(min, max, ave): 0.050s, 6.606s, 1.047s

number of prompt tokens: 244330
number of completion tokens: 240421
token throughput (completion token): 1517.721 token/s
token throughput (prompt + completion token): 3060.118 token/s
RPS (request per second): 6.313 req/s
RPM (request per minute): 378.766 req/min
--------------------------------------------------
```


reproduce step:
1. convert to get turbomind model 
`lmdeploy convert llama2 /mnt/bigdisk/qa_test_models/meta-llama/Llama-2-7b-chat-hf  --dst-path ./workspace
`2. start triton server
`bash workspace/service_docker_up.sh
`
3. do benchmark

```
python3 benchmark/profile_serving.py localhost:33337 workspace/triton_models/tokenizer/tokenizer.model ../ShareGPT_V3_unfiltered_cleaned_split.json --concurrency 64
python3 benchmark/profile_serving.py localhost:33337 workspace/triton_models/tokenizer/tokenizer.model ../ShareGPT_V3_unfiltered_cleaned_split.json --concurrency 128

```




**2. triton python backend with hf model files**
```
--------------------------------------------------
concurrency: 64
elapsed_time: 99.006s

first_token latency(min, max, ave): 0.025s, 1.062s, 0.090s

number of prompt tokens: 244330
number of completion tokens: 239421
token throughput (completion token): 2418.255 token/s
token throughput (prompt + completion token): 4886.093 token/s
RPS (request per second): 10.100 req/s
RPM (request per minute): 606.026 req/min
--------------------------------------------------

--------------------------------------------------
concurrency: 128
elapsed_time: 85.037s

first_token latency(min, max, ave): 0.042s, 1.770s, 0.228s

number of prompt tokens: 244330
number of completion tokens: 239421
token throughput (completion token): 2815.508 token/s
token throughput (prompt + completion token): 5688.743 token/s
RPS (request per second): 11.760 req/s
RPM (request per minute): 705.579 req/min
--------------------------------------------------
```

reproduce step:
1. accord to readme to create model_repository folder
2. copy hf model files into weights
3. start triton server, use service_docker_up.sh make sure in the same docker images. 
`bash service_docker_up.sh`
4. do benchmark
```
python3 benchmark/profile_triton_python_backend.py localhost:33337 /path-to/meta-llama/Llama-2-7b-chat-hf ../ShareGPT_V3_unfiltered_cleaned_split.json --num_prompts 1000 --concurrency=128

python3 benchmark/profile_triton_python_backend.py localhost:33337 /path-to/meta-llama/Llama-2-7b-chat-hf ../ShareGPT_V3_unfiltered_cleaned_split.json --num_prompts 1000 --concurrency 64
```


**3. triton python backend with turbomind model files**
```

--------------------------------------------------
concurrency: 64
elapsed_time: 84.353s

first_token latency(min, max, ave): 0.025s, 1.483s, 0.114s

number of prompt tokens: 220954
number of completion tokens: 213092
token throughput (completion token): 2526.188 token/s
token throughput (prompt + completion token): 5145.580 token/s
RPS (request per second): 11.855 req/s
RPM (request per minute): 711.295 req/min
--------------------------------------------------
--------------------------------------------------
concurrency: 128
elapsed_time: 72.332s

first_token latency(min, max, ave): 0.040s, 2.367s, 0.258s

number of prompt tokens: 220954
number of completion tokens: 213092
token throughput (completion token): 2946.030 token/s
token throughput (prompt + completion token): 6000.753 token/s
RPS (request per second): 13.825 req/s
RPM (request per minute): 829.509 req/min
--------------------------------------------------
```

reproduce step:
1. accord to readme to create model_repository folder
2. convert to get turbomind model 
`lmdeploy convert llama2 /mnt/bigdisk/qa_test_models/meta-llama/Llama-2-7b-chat-hf  --dst-path ./workspace
`
3. copy workspace model files into weights
4. start triton server, use service_docker_up.sh make sure in the same docker images. 
`bash service_docker_up.sh
`
5. do benchmark
```
python3 benchmark/profile_triton_python_backend.py localhost:33337 /path-to/meta-llama/Llama-2-7b-chat-hf ../ShareGPT_V3_unfiltered_cleaned_split.json --num_prompts 1000 --concurrency=128

python3 benchmark/profile_triton_python_backend.py localhost:33337 /path-to/meta-llama/Llama-2-7b-chat-hf ../ShareGPT_V3_unfiltered_cleaned_split.json --num_prompts 1000 --concurrency 64
```

**Comment by zhulinJulia24** [comment]
> Hi @zhulinJulia24 Under normal circumstances, the performance should be flat or slightly better compared to the API Server, and there should not be a doubling situation. Perhaps you can share your reproduction steps so we can verify it locally.

By using hf models with new triton python backend, the performance is the best.

**Comment by zhyncs** [comment]
> > Hi @zhulinJulia24 Under normal circumstances, the performance should be flat or slightly better compared to the API Server, and there should not be a doubling situation. Perhaps you can share your reproduction steps so we can verify it locally.
> 
> By using hf models with new triton python backend, the performance is the best.

@ispobock please follow up

**Comment by ispobock** [comment]
@zhulinJulia24 In my testing, the performance is the same for Turbomind workspace and hf model. Could you test the performance for api_server with your converted Turbomind workspace?

**Comment by zhulinJulia24** [comment]
> @zhulinJulia24 In my testing, the performance is the same for Turbomind workspace and hf model. Could you test the performance for api_server with your converted Turbomind workspace?

I did it on 0.2.6 version
concurrency = 128
token throughput (completion token): 3020.925333 
rpm: 778.8856667 

concurrency = 64
token throughput (completion token): 2591.198
rpm: 668.0893333

**Comment by zhulinJulia24** [comment]
> @zhulinJulia24 In my testing, the performance is the same for Turbomind workspace and hf model. Could you test the performance for api_server with your converted Turbomind workspace?

It seems I put  internlm2 benchmark on it, I'm not sure. 
I did again, the resutl is updated. Can you help to have a look?

**Comment by zhulinJulia24** [comment]
> > @zhulinJulia24 In my testing, the performance is the same for Turbomind workspace and hf model. Could you test the performance for api_server with your converted Turbomind workspace?
> 
> It seems I put internlm2 benchmark on it, I'm not sure. I did again, the resutl is updated. Can you help to have a look?

@ispobock   the performance of Turbomind workspace and hf model with new triton python is similar. Compared to old triton server, perfomance doubled.   Does this result meet your expectations?

**Comment by ispobock** [comment]
@zhulinJulia24 Yes, performance of Turbomind workspace and hf model should be similar. 

My previous test setting:
- model: `llama2-13b-chat`
- tp: `1`
- concurrency: `128`

Test result:
- Old C++ Triton backend (with 128 instance group count): `5.9` RPS
- This Python Triton backend: `6.7` RPS (for both Turbomind workspace and hf model)
- api_server: `6.7` RPS (for both Turbomind workspace and hf model)

**Review by zhulinJulia24 — APPROVED** [review]
lgtm

File: lmdeploy/serve/turbomind/triton_python_backend/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# LMDeploy Triton Python Backend
+
+We provide the [Triton Python Backend](https://github.com/triton-inference-server/python_backend/tree/main) for LMDeploy. Here is the guide to deploy server, test client and profile server performance.
+
+## Server Deployment
+
+### Step 1: Prepare the Model Repository
+
+The Triton Inference Server requires a specific folder structure for serving models. We provided the necessary `model.py` and `config.pbtxt` files. Follow these steps to set up your model repository:
```
Comment (zhyncs) [inline lmdeploy/serve/turbomind/triton_python_backend/README.md:9]: Please add the configuration instructions for `tp` in config.pbtxt.

Comment (ispobock) [reply lmdeploy/serve/turbomind/triton_python_backend/README.md:9]: updated

**Comment by zhyncs** [comment]
LGTM

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**